### PR TITLE
[gf] Chebyshev mesh + refactoring

### PR DESCRIPTION
**Pull request by iskakoff**

### CODE REVIEW

**Comment by egull** [comment]
Can you fix up the MPI builds first?

File: gf/include/alps/gf/mesh/chebyshev_mesh.hpp [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        double z = cos((2 * i + 1) * M_PI / (2 * k_));
+        zeros_[i] = z * beta_ / 2. + beta_ / 2.;
+      }
+
+    }
+  };
+
+  ///Stream output operator, e.g. for printing to file
+  std::ostream &operator<<(std::ostream &os, const chebyshev_mesh &M);
+}}
```
Comment (galexv) [inline gf/include/alps/gf/mesh/chebyshev_mesh.hpp]: No EOL before EOF. Not a big deal, but funny enough [it's UB in C++03](https://stackoverflow.com/a/72450) (but not in C++11). No need to do anything now, but we may ask contributors to ensure there is EOL at the last line.

File: gf/include/alps/gf/mesh/numerical_mesh.hpp [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    statistics::statistics_type statistics() const{ return statistics_;}
+    const piecewise_polynomial<T>& basis_function(int l) const {
+      assert(l>=0 && l < dim_);
+      check_validity();
+      return basis_functions_[l];
+    }
+
+
+    /// Swaps this and another mesh
+    // It's a member function to avoid dealing with templated friend decalration.
```
Comment (egull) [inline gf/include/alps/gf/mesh/numerical_mesh.hpp]: declaration

File: gf/include/alps/gf/mesh/numerical_mesh.hpp [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+#include <alps/gf/mpi_bcast.hpp>
+#endif
+#include <alps/gf/piecewise_polynomial.hpp>
+#include <alps/gf/mesh/index.hpp>
+#include <alps/gf/mesh/mesh_base.hpp>
+
+namespace alps {namespace gf {
+
+  /**
+   * Numerical mesh (T = double or std::complex<double>)
```
Comment (egull) [inline gf/include/alps/gf/mesh/numerical_mesh.hpp]: what is a 'numerical' mesh? Description?

File: gf/include/alps/gf/mesh/frequency_meches.hpp [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+      if (mesh::frequency_positivity_type(pos)!=positivity_) {
+        throw std::invalid_argument("Attempt to broadcast Matsubara mesh with the wrong positivity type "+std::to_string(pos) ); // FIXME: specific exception? Verbose positivity?
+      };
+      offset_ = ((PTYPE==mesh::POSITIVE_ONLY)?0:nfreq_/2);
+
+      try {
+        check_range();
+      } catch (const std::exception& exc) {
+        int wrank=alps::mpi::communicator().rank();
+        // FIXME? Try to communiucate the error with all ranks, at least in debug mode?
```
Comment (egull) [inline gf/include/alps/gf/mesh/frequency_meches.hpp]: communicate

File: gf/include/alps/gf/mesh/chebyshev_mesh.hpp [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    alps::gf::statistics::statistics_type statistics_;
+
+    ///zeros of the chebyshev polynomial of order k
+    std::vector<double> zeros_;
+
+    inline void throw_if_empty() const {
+      if (extent() == 0) {
+        throw std::runtime_error("chebyshev mesh is empty");
+      }
+    }
```
Comment (mwallerb) [inline gf/include/alps/gf/mesh/chebyshev_mesh.hpp]: Suggest surrounding `throw_if_empty` with NDEBUG guard:
```C++
    inline void throw_if_empty() const {
#ifdef NDEBUG
      if (extent() == 0)
        throw std::runtime_error("chebyshev mesh is empty");
#endif 
    }
```

File: gf/include/alps/gf/mesh/chebyshev_mesh.hpp [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+    chebyshev_mesh(double beta, int k) : beta_(beta), k_(k), statistics_(alps::gf::statistics::FERMIONIC) {
+      compute_zeros();
+    }
+
+    int operator()(index_type idx) const {
+#ifndef NDEBUG
+      throw_if_empty();
+#endif
```
Comment (mwallerb) [inline gf/include/alps/gf/mesh/chebyshev_mesh.hpp]: Propose to remove this in favor of guard in `throw_if_empty()`

File: gf/include/alps/gf/mesh/frequency_meches.hpp (hunk 4 above)
Comment (mwallerb) [inline gf/include/alps/gf/mesh/frequency_meches.hpp]:
```suggestion
        // FIXME? Try to communicate the error with all ranks, at least in debug mode?
```

File: gf/include/alps/gf/mesh/numerical_mesh.hpp (hunk 2 above)
Comment (mwallerb) [inline gf/include/alps/gf/mesh/numerical_mesh.hpp]:
```suggestion
    // It's a member function to avoid dealing with templated friend declaration.
```

**Comment by egull** [comment]
Thank you very much Sergei!

**Pull request merged**